### ISSUE TITLE
Need a command for checking readiness/liveness

### ISSUE THREAD

**Issue by alchemistake** [taskiq-python/taskiq#240]
I think it will really help if there is command like these:
`$ taskiq list workers`
```
Worker name | State
worker1 Listening
worker2 Closing
worker3 Starting
```
----
`$ taskiq status worker1`
`Listening`

so i can write a readiness probe for kubernetes using these commands.

**Comment by s3rius** [comment]
Yes. Thanks for your interest in the project. This is a great idea. We will try to implement when possible.

**Comment by itssimon** [comment]
Another use case for this are Kubernetes readiness and liveness checks for a particular worker pod.

**Comment by WiZeYAR** [comment]
@s3rius, @alchemistake, should we continue this discussion?
To me it doesn't seem as a hard to implement thing. We could embed a basic http listen like this:

```python
import asyncio

async def handle_request(reader, writer):
    request = await reader.read(100)
    print(request.decode())
    
    response = b"HTTP/1.1 200 OK\r\nContent-Type: text/plain\r\n\r\nReady to accept connections!"
    writer.write(response)
    await writer.drain()
    
    writer.close()

async def main():
    server = await asyncio.start_server(handle_request, '0.0.0.0', 8000)
    
    addr = server.sockets[0].getsockname()
    print(f'Serving on {addr}')
    
    async with server:
        await server.serve_forever()

if __name__ == "__main__":
    asyncio.run(main())
```

The solution won't need any external deps, and it will only require us to make a hook to the startup logic.

**Comment by WiZeYAR** [comment]
We can add a `/healthz` route as an option to the worker/scheduler. The use will have to specify an http listen port.

**Comment by s3rius** [comment]
I was thinking about this approach, but we need to design the solution a bit more carefully.

I believe this can be useful for two purposes:

1. Health checks
2. Collecting debugging information from worker processes

To make the health checks more accurate, it would be great if we could verify the connection liveness directly inside the worker. This got me thinking that we could start servers in each worker process and have a central server in the process manager that can query any worker process.

This would give us better error detection capabilities and allow us to gather aggregated information from each process. However, this approach requires some effort in determining which port to open from the master process.

We could start by binding ports in the process manager using the `SO_REUSEPORT` flag, and then pass the allocated port to the child processes. When a worker process restarts, we can reuse the same port, which is a nice property. But I'm wondering if there's a better solution.

For instance, we could bind to port `0` (which will allocate any available port) on the workers and use something like [multiprocessing.Value](https://docs.python.org/3/library/multiprocessing.html#multiprocessing.Value) to communicate back to the process manager, letting it know which port has been bound.

**Comment by WiZeYAR** [comment]
@s3rius, what kind of debug info should the workers report?
We can add another async procedure or extend the liveness protocol I've drafted above.

**Comment by s3rius** [comment]
I think we should use NET sockets, instead of UNIX, because as I know, they don't work properly on windows. 

Also, heartbeats can be quite different. Liveness of the worker process can be easily obtained from the main process. Because it has all the pids. But I guess we want to notify about problems with current connection to the queue or something similar. And your design might work quite good for that purpose.  

In regards to debug info, I guess we can just send initial request when worker gets up, right after startup and task discovering we will send one request with all the details we have collected so far.  It includes:
* Name of tasks
* Path to the broker
* All the labeled schedules if any
* Maybe current broker's internal state.

**Comment by benedikt-bartscher** [comment]
Windows is useless

**Comment by s3rius** [comment]
Agree, but we still need to support it.

**Comment by WiZeYAR** [comment]
I now think, whether we should split readiness/liveness and telemetry (debug log) into separate issues. For debug logging, we could use something like Open Telemetry, as this is a standard for doing logs, and it would allow taskiq to be integrated into existing systems (and be observable).

> Liveness of the worker process can be easily obtained from the main process.

Does this also cover the cases when the worker got stuck executing e.g. [`polars`](https://pola.rs) code (i.e. executing native code)? If it is so, then checking on the subprocess is the best option. I wonder if we can test cases like that...

Still, the simpler it is -- the better.

**Comment by s3rius** [comment]
> Does this also cover the cases when the worker got stuck executing
I guess yes, because even if you're going to execute sync code, taskiq uses Executors to execute this code on processpool or threadpool. And therefore the loop and the main task for listening to messages won't be blocked.

> Still, the simpler it is -- the better.

Absolutely.

We can make it even simpler, by just writing statuses to a file periodically. The file will be located at `/tmp/taskiq_status_{pid}.txt`. And removed at the end of the session.

Once every second or every change in processes we will be updating this file. We can make a small CLI command like `taskiq health` that reads this file. 


Or alternatively we can write statuses of each process and their connection status separately from each other. The main process will be collecting information from their files and just displaying them.

**Comment by lryan599** [comment]
This feature is important, and I hope to see it sooner.

**Comment by WiZeYAR** [comment]
I'll start studying the codebase tomorrow

**Comment by WiZeYAR** [comment]
I made a fork, and did some code reading this weekend.

**My observations:**
1. You have a good middleware system, capable of a lot of things
2. The middleware seems to only affect workers' behavior, but not the manager's

**My questions:**
1. If the observations of mine are correct, then how does `PrometheusMiddleware` work? Assuming you have 4 worker processes, each trying to allocate port 9000 -- do 3 of these workers just post a debug message `Cannot start prometheus server`?

**Options I see:**
1. Make middleware also affect managers (maybe make some other kind of middleware for them)
    > The implementation might be trickier, but it should be future-proof, as I have a feeling that this project will be making adjustments to the manager in the future
2. Go middleware way, having `SO_REUSEPORT`, as was mentioned above (does it also work in Windows?)
3. Hardcode liveness on manager side, having middleware on the worker side
4. Hardcode liveness everywhere

**Thing to decide, if we go any option other than 2:**
Where does the communication between manager and worker lie? We did a discussion before, but this is @s3rius decision (nickname fits here 😆), as he's the maintainer.

We have the following communication means:
1. Tempfiles
2. Net socket
3. ~~Unix socket~~ -- fastest way, but won't work everywhere

I'm ready to start the implementation myself, if blessing is given.

I might need your input here, @s3rius

**Comment by WiZeYAR** [comment]
Any thoughts, @s3rius?

**Comment by corruptmane** [comment]
bump

**Comment by WiZeYAR** [comment]
Hey, @corruptmane, @alchemistake, as I have said, there's draft impl for this: https://github.com/taskiq-python/taskiq/pull/489

Can we somehow collaborate on its testing, if any of you are kube guys?
